gnomad/utils/constraint.py

```diff
@@ -877,3 +839,344 @@ def compute_expected_variants(
+    # and expected variants and compute observed:expected ratio.
+    agg_expr = {
+        "obs": hl.agg.sum(ht.observed_variants),
+        "exp": hl.agg.sum(ht.expected_variants),
+        "possible": hl.agg.sum(ht.possible_variants),
+    }
+    agg_expr["oe"] = agg_expr["obs"] / agg_expr["exp"]
+
+    # Create an aggregator that sums the mutation rate.
+    if not exclude_mu_sum:
+        agg_expr["mu"] = hl.agg.sum(ht.mu)
+
+    # Create aggregators that sum the number of observed variants
+    # and expected variants for each population if pops is specified.
+    if pops:
+        agg_expr["pop_exp"] = hl.struct(
+            **{pop: hl.agg.array_sum(ht[f"expected_variants_{pop}"]) for pop in pops}
+        )
+        agg_expr["pop_obs"] = hl.struct(
+            **{pop: hl.agg.array_sum(ht[f"downsampling_counts_{pop}"]) for pop in pops}
+        )
+
+    # TODO: ask hail team why the agg.filter statement returns an error.
+    # return hl.agg.filter(filter_expr, hl.struct(**agg_expr))
+    return hl.agg.group_by(filter_expr, hl.struct(**agg_expr)).get(True)
+
+
+def get_all_pop_lengths(
+    ht: hl.Table,
+    pops: Tuple[str],
+    obs_expr: hl.expr.StructExpression,
+) -> List[Tuple[str, str]]:
+    """
+    Get the minimum length of observed variant counts array for each population downsampling.
+
+    The observed variant counts for each population in `pops` are specified by
+    annotations on the `obs_expr` expression.
+
+    The function also performs a check that arrays of variant counts within population
+    downsamplings all have the same lengths.
+
+    :param ht: Input Table containing `obs_expr`.
+    :param pops: Populations used to categorize observed variant counts in downsamplings.
+    :param obs_expr: Expression for the population observed variant counts. Should be a
+        struct containing an array for each pop in `pops`.
+    :return: A Dictionary with the minimum array length for each population.
+    """
+    # TODO: This function will be converted into doing just the length check if there
+    #  is no usage of pop_lengths in the constraint pipeline.
+    # Get minimum length of downsamplings for each population.
+    pop_downsampling_lengths = ht.aggregate(
+        [hl.agg.min(hl.len(obs_expr[pop])) for pop in pops]
+    )
+
+    # Zip population name with their downsampling length.
+    pop_lengths = list(zip(pop_downsampling_lengths, pops))
+    logger.info("Found: %s", "".join(map(str, pop_lengths)))
+
+    assert ht.all(
+        hl.all(
+            lambda f: f,
+            [hl.len(obs_expr[pop]) == length for length, pop in pop_lengths],
+        )
+    ), (
+        "The arrays of variant counts within population downsamplings have different"
+        " lengths!"
+    )
+
+    return pop_lengths
+
+
+def compute_pli(
+    ht: hl.Table,
+    obs_expr: hl.expr.Int32Expression,
+    exp_expr: hl.expr.Float32Expression,
+    expected_values: Dict[str, float] = {"Null": 1.0, "Rec": 0.463, "LI": 0.089},
```

Comment:
Changed this to None and then use these defaults if it's set to None to handle mutable default warning